// ts-ignore/lib/actions/getVar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = getVar;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _services = require("../services");

const _process = process,
      env = _process.env;

function getVar(_x, _x2, _x3) {
  return _getVar.apply(this, arguments);
}

function _getVar() {
  _getVar = (0, _asyncToGenerator2.default)(function* (config, logger, key) {
    const git = new _services.Git(config.accounts);
    if (!key) return null;

    const _ref = yield git.getAccount(env.CAPTAIN_HOOK_STRATEGY || 'random'),
          email = _ref.email,
          name = _ref.name;

    const result = (key => ({
      git_author_date: 'unknown',
      git_author_email: email,
      git_author_name: name,
      git_committer_date: 'unknown',
      git_committer_email: email,
      git_committer_name: name
    })[key])(key);

    return logger.info(result);
  });
  return _getVar.apply(this, arguments);
}
