// ts-ignore/lib/services/git.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _asyncCrossSpawn = _interopRequireDefault(require("async-cross-spawn"));

var _sample = _interopRequireDefault(require("lodash/sample"));

class Git {
  constructor(accounts = []) {
    this.accounts = accounts;
  }

  getAccount(strategy = 'config') {
    var _this = this;

    return function () {
      var _ref = (0, _asyncToGenerator2.default)(function* (strategy) {
        return {
          config: {
            name: yield (0, _asyncCrossSpawn.default)('git', ['config', '--global', 'user.name']),
            email: yield (0, _asyncCrossSpawn.default)('git', ['config', '--global', 'user.email'])
          },
          random: (0, _sample.default)(_this.accounts)
        }[strategy];
      });

      return function (_x) {
        return _ref.apply(this, arguments);
      };
    }()(strategy);
  }

}

exports.default = Git;
