// ts-ignore/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.symbol.description");

require("core-js/modules/es.array.iterator");

require("core-js/modules/es.array.map");

require("core-js/modules/es.array.splice");

require("core-js/modules/es.object.values");

require("core-js/modules/es.promise");

require("core-js/modules/es.string.match");

require("core-js/modules/es.string.split");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _execa = _interopRequireDefault(require("execa"));

var _runtime = require("babel-plugin-ts-optchain/lib/runtime");

var _fsExtra = _interopRequireDefault(require("fs-extra"));

var _ora = _interopRequireDefault(require("ora"));

var _path = _interopRequireDefault(require("path"));

var _pkgDir = _interopRequireDefault(require("pkg-dir"));

var _command = require("@oclif/command");

var _bluebird = require("bluebird");

const rootPath = _pkgDir.default.sync(process.cwd()) || process.cwd();

class TSIgnore extends _command.Command {
  run() {
    return (0, _asyncToGenerator2.default)(function* () {
      const spinner = (0, _ora.default)();
      spinner.start('finding errors');
      const files = {};
      const lines = (yield (0, _execa.default)('tsc', ['--noEmit'], {
        cwd: rootPath
      }).catch(err => {
        if (err.stdout) return err;
        throw err;
      })).stdout.split('\n');
      spinner.start('fixing errors');
      let count = 0;
      yield (0, _bluebird.mapSeries)(lines,
      /*#__PURE__*/
      function () {
        var _ref = (0, _asyncToGenerator2.default)(function* (line) {
          const _ref2 = line.match(/(.+\.tsx?)\((\d+),\d+\): error TS\d{4}: /) || [null, null, null],
                _ref3 = (0, _slicedToArray2.default)(_ref2, 3),
                filePath = _ref3[1],
                lineNumber = _ref3[2];

          if (!filePath) return;

          if (!(filePath in files)) {
            files[filePath] = {
              count: 0,
              lines: (yield _fsExtra.default.readFile(filePath)).toString().split('\n'),
              path: _path.default.resolve(process.cwd(), filePath)
            };
          }

          ++count;
          const file = files[filePath];
          const padding = (0, _runtime.oc)(file.lines[Number(lineNumber) + file.count - 1].match(/^ */), [], [''])[0].length;
          file.lines.splice(Number(lineNumber) + file.count - 1, 0, `${Array(padding).fill(' ').join('')}// @ts-ignore`);
          ++file.count;
        });

        return function (_x) {
          return _ref.apply(this, arguments);
        };
      }());
      yield Promise.all(Object.values(files).map(
      /*#__PURE__*/
      function () {
        var _ref4 = (0, _asyncToGenerator2.default)(function* (file) {
          yield _fsExtra.default.writeFile(file.path, file.lines.join('\n'));
        });

        return function (_x2) {
          return _ref4.apply(this, arguments);
        };
      }()));
      spinner.succeed(`fixed ${count} errors`);
    })();
  }

}

exports.default = TSIgnore;
TSIgnore.description = 'ignore typescript errors';
TSIgnore.flags = {
  help: _command.flags.help({
    char: 'h'
  }),
  version: _command.flags.version({
    char: 'v'
  })
};
